### ISSUE TITLE
Ability to play an audio buffer from an extension

### ISSUE THREAD

**Issue by jaustin** [microsoft/pxt-microbit#4609]
**Is your feature request related to a problem? Please describe.**

When developing sound-related features for MakeCode, we'd like to prototype things as extensions.
At the moment there's no good way for us to get equivalent audio on the simulator to what we can generate on the micro:bit device with extensions (either C++ or typescript).

This applies to both the work @CaptainCredible is doing on synths and @JohnVidler is doing on record and playback

**Describe the solution you'd like**

I *think* we'd like to be able give the simulator a PCM-like buffer that it will play as audio

* In the case of the synth, the synth can be re-implemented in TS to generate appropriate buffers when running on the sim versus running on the device when we need a C++ codepath
* In the case of the recording and playback API, we can temporarily use simple pre-recorded buffers to represent the user's recorded audio (for example, a pre-recorded set of beeps or something that gives at least minimal audio feedback that something happened.

**Describe alternatives you've considered**
For the record/playback case it would be preferable in the long term for the simulator to be able to record and playback sound in line with the new CODAL APIs, but I understand this isn't achievable in the short term. @JohnVidler could you please point at the APIs for context?

**Additional context**

There's at least two strands of work where this is useful, so a simulator/audio expert like @riknoll would likely be better placed to know whether this is actually what we really want, versus some other way for extensions to do audio simulation.

**Comment by riknoll** [comment]
I tried an experiment with this a while ago, and unfortunately our simulator just isn't fast enough to run at the audio rates required to generate PCM buffers.

The best alternative would be to surface webaudio APIs in the simulator through a series of shim functions and then using those to approximate the same output as the C++. I've been thinking about doing something like this for a while...

**Comment by jaustin** [comment]
Some of what we're interested here will be pre-determined buffers (eg audio clips) so the generation rate isn't so significant.

@riknoll what's being generated for the sound effects right now? I think I'd guessed that was generating PCM?

**Comment by riknoll** [comment]
yes it is generating PCM. the difference between the two scenarios is that the current audio generating code lives in the "C++" layer as opposed to the user code layer, which means it doesn't go through our compiler and is significantly faster in the sim.

if we want to do an implementation of the synth, it would have to live inside pxt-core or pxt-common-packages and not the extension itself. we don't allow extensions to load their own "C++" javascript code for security reasons.

**Comment by jaustin** [comment]
So playing pre-made buffers would be likely viable, but composing buffers not? I think being able to fling a buffer at the existing infrastructure would be really useful (consider, say, a bird sounds from around the world extension, which is mainly samples)
